core/src/io/github/lonamiwebs/klooni/game/State.java

```diff
@@ -0,0 +1,282 @@
+                else
+                    emptySpace += 1;
+    }
+
+    // endregion
+
+    //region Check piece
+
+    private static String getUnfitBlock() {
+        StringBuilder sb = new StringBuilder();
+        for (int i : unfitBlock) {
+            sb.append(i);
+            sb.append(" ");
+        }
+        return sb.toString();
+    }
+
+    // True if the given cell coordinates are inside the bounds of the board
+    private boolean inBounds(int x, int y) {
+        return x >= 0 && x < cellCount && y >= 0 && y < cellCount;
+    }
+
+    // True if the given piece at the given coordinates is not outside the bounds of the board
+    private boolean inBounds(Piece piece, int x, int y) {
+        return inBounds(x, y) && inBounds(x + piece.cellCols - 1, y + piece.cellRows - 1);
+    }
+
+    // Given coordinates as the starting point, return true if a piece fits in said coordinates
+    private boolean canPutPiece(Piece piece, int x, int y) {
+        if (!inBounds(piece, x, y))
+            return false;
+
+        for (int i = 0; i < piece.cellRows; ++i)
+            for (int j = 0; j < piece.cellCols; ++j)
+                if (state[y + i][x + j] && piece.filled(i, j)) {
+                    return false;
+                }
+
+        return true;
+    }
+
+    // Given Vector2 as last checked coordinates, return a new vector where the piece fits,
+    // else the same vector if there are no longer any place in board to fit the piece
+    private Vector2 canPutPiece(Piece piece, Vector2 origin) {
+        for (int j = (int) (origin.x + 1); j < cellCount; ++j)
+            if (canPutPiece(piece, j, (int) origin.y))
+                return new Vector2(j, origin.y);
+
+        for (int i = (int) (origin.y + 1); i < cellCount; ++i)
+            for (int j = 0; j < cellCount; ++j)
+                if (canPutPiece(piece, j, i))
+                    return new Vector2(j, i);
+
+        return origin;
+    }
+
+    // Check every possible state from a set of blocks and an initial state.
+    // Immediately return true if found one fitting occurrence.
+    private static boolean checkPermute(Piece[] holder, State state) {
+        Vector2 temp1, temp2, temp3, pos1 = ORIGIN, pos2 = ORIGIN;
+        for (int i = 0; i < 6; i++) {
+            unfitBlock[i] = -1;
+            // TODO: Optimize the checking algorithm (use any pruning techniques?)
+            // TODO: Determine better criteria to change unfit block
+            while(true) {
+                State state1 = new State(state);
+                temp1 = state.canPutPiece(holder[CHECKER[i][0]], pos1);
+                if (temp1.epsilonEquals(pos1, MathUtils.FLOAT_ROUNDING_ERROR)) {
+                    if (unfitBlock[i] < 0)
+                        unfitBlock[i] = 0;
+                    break;
+                }
+                state1.putPiece(holder[CHECKER[i][0]], temp1);
+                while (true) {
+                    State state2 = new State(state1);
```

Comment:
If I do `state2 = state1`, Java would only shallow copy the reference, so when I change `state2` it'll change `state1` too. I need deep copy, and the way to implement it in Java is by create an entire new object containing the exact same values.

The infinite loop is used because we have no idea how many possible position to put a block, so we'll just check all possible position until we find no more--that's when we break the loop.

Reply:
Yes, I know how references work. Just pointing out that this algorithm is expensive to execute :P but the board isn't *that* big (it's just 100 tiles) so it should be alright. I have not ran the code yet.